### ISSUE TITLE
`BlockSparseAttention` doesn't use dropout

### ISSUE THREAD

**Issue by fmassa** [facebookresearch/xformers#122]
# 🐛 Bug

While working on refactoring the `blocksparse.py` file into its own custom tensor type, I noticed that `BlockSparseAttention` doesn't implement dropout, which is probably not intended.

cc @blefaudeux

**Comment by blefaudeux** [comment]
Wow not intended indeed, should be easy enough to fix but oversight :/ thanks for spotting this

**Comment by blefaudeux** [comment]
I just had a look, it's not trivially supported right now, we should not expose it as a setting at the very least. A vanilla dropout will not do since at that point the attention matrix is sparse, we could change our [triton based](https://github.com/facebookresearch/xformers/blob/main/xformers/triton/dropout.py) dropout to handle block sparse matrices, some work but doable. Another option is to dropout using [a random mask on the attention](https://github.com/facebookresearch/xformers/blob/main/xformers/components/attention/blocksparse.py#L187), this should be well handled top of mind, it's probably a more elegant take

**Comment by blefaudeux** [comment]
Not correct dimension wise (using the same mask across heads and batch), but quick [example for the second take](https://github.com/facebookresearch/xformers/blob/blocksparse_dropout_fix/xformers/components/attention/blocksparse.py#L187)

**Comment by fmassa** [comment]
@blefaudeux can't we call PyTorch's dropout in the output of the softmax? IIUC Triton's kernel is just taking and returning a dense Tensor (which corresponds to the location where the elements are non-sparse)?

**Comment by fmassa** [comment]
(this is btw what we do in https://github.com/facebookresearch/xformers/blob/71bab94cb954e6e291ca93d3bce5dffadab4286d/xformers/components/attention/core.py#L159-L174 )

**Comment by blefaudeux** [comment]
> @blefaudeux can't we call PyTorch's dropout in the output of the softmax? IIUC Triton's kernel is just taking and returning a dense Tensor (which corresponds to the location where the elements are non-sparse)?

top of mind I think that the output of the softmax is blocksparse, so torch dropout will not work on that directly.

**Comment by fmassa** [comment]
The output of `softmax` is just a standard `torch.Tensor` of the non-zero values. https://github.com/openai/triton/blob/b7f0e87dc2143cdf9cbc050e34ba80440074c53b/python/triton/ops/blocksparse/matmul.py#L233

**Comment by blefaudeux** [comment]
> The output of `softmax` is just a standard `torch.Tensor` of the non-zero values. https://github.com/openai/triton/blob/b7f0e87dc2143cdf9cbc050e34ba80440074c53b/python/triton/ops/blocksparse/matmul.py#L233

Ok then perfect, that works and it's a one liner !
